Augmented Bond Curves

JakeHartnell opened the pull request:
ABCs are cool! This will try and implement much of what is in this [excellent blog post](https://medium.com/commonsstack/deep-dive-augmented-bonding-curves-b5ca4fad4436).

To do:
- [x] Add max supply option
- [x] Update curve
- [x] Update max supply
- [x] Close bonding curve (only sell)
- [x] MinMax contribution amounts for hatching
- [x] Fix token metadata so that it's set correctly
- [x] Test allowlist
- [x] Test close curve
- [x] Test update curve
- [x] Test update max supply
- [x] Factory contract for creating cw-abc DAOs
- [ ] Curve type validation
- [ ] Move related contracts to `experimental` folder

Bonus:
- [ ] Move curves into a new `cw-curves` package
- [ ] Add more curves (Taylor Series. S-Curve, Piecewise Linear)
- [ ] Extend allowlist options (i.e. keep existing list, but add a query to check if an address is a member of a DAO)

JakeHartnell committed:
Initial commit

bmorphism commented:
doing G-d's work, Ser! 👏

adairrr committed:
Initial phase integration to cw-abc (#698)

* CwAbcResult

* Hatch phase configuration and init msg refactor

* Initial instantiate test

* Implement separate phrases and phrase configs

* Remove vesting phase and update funding pool

* Separate commands and queries

* Update init msg with string configs

* Phase config query

* MinMax and config query

* Add some todos

* cw-ownable integration

adairrr committed:
cw-abc: Updated hatch phase mechanics, donations, queries (#699)

* Separate hatcher allowlist

* Donation feature

* Initial sell exit tax

* Hatchers to amount

* Hatch phase exit tax

* TokenMsg methods

* Format

* Hatchers query

* Fix bug where float was not taken into account in supply

* Buy and sell refactoring

* Update hatch phase config

* Update phase config enum

* Add adairrr to authors

* Initial boot integration with custom msgs

* Initial testing infrastructure

* Abstract-OS to AbstractSDK

JakeHartnell committed:
Comments clean up

Dat-Andre committed:
697 (#702)

* update cw-orch dependency

* add getrandom dependency to .toml

---------

Co-authored-by: Jake Hartnell <Jake.Hartnell@gmail.com>

JakeHartnell committed:
Remove boot / cw-orch, get cw-multi-test working with custom messages

JakeHartnell committed:
Attempt to get tests working

JakeHartnell committed:
Start prepping for refactor and tests

JakeHartnell committed:
Refactor cw-abc contract to work with cw-tokenfactory-issuer

JakeHartnell committed:
Remove old attempt at fixing cw-multi-test

JakeHartnell committed:
Refactor

JakeHartnell committed:
Update schema

JakeHartnell committed:
Clean up and notes

JakeHartnell committed:
Get test-tube tests running

JakeHartnell committed:
Get instantiation working with test-tube

JakeHartnell committed:
Fix tests, need to investigate why this broke

JakeHartnell committed:
Minting tokens works!

JakeHartnell committed:
Fixups, burning tokens still doesn't work

JakeHartnell committed:
Rebase cleanup, fix up tests

JakeHartnell committed:
More cleanup for clippy

JakeHartnell committed:
Attempt to fix burn

JakeHartnell committed:
Note

JakeHartnell committed:
Clean up, clippy, workspace deps, notes

JakeHartnell committed:
Happy path test, make notes

JakeHartnell committed:
Refactor update phase config, fix metadata, clean up

JakeHartnell committed:
Contribution limits

JakeHartnell committed:
Max Supply

JakeHartnell committed:
Clean up, add notes for future work

JakeHartnell committed:
Close curve, update curve, set max supply methods

JakeHartnell committed:
Rename methods, cleanup unused variables

JakeHartnell committed:
Update schema

JakeHartnell committed:
Fix some tests

JakeHartnell committed:
More tests!

JakeHartnell committed:
Attempt at update curve tests

JakeHartnell committed:
Next pass at adding info to readme

JakeHartnell committed:
to_binary -> to_json_binary

JakeHartnell committed:
Bug fixes, implement fees

JakeHartnell committed:
Fix exit fees

JakeHartnell committed:
Update schema

JakeHartnell committed:
Add more information to the readme

JakeHartnell committed:
Rename variables to make them more consistent and understandable

JakeHartnell committed:
More queries, update schema

JakeHartnell committed:
Remove TokenFactoryMsg and TokenFactoryQuery, no token-bindings dep

JakeHartnell committed:
Remove unused old tests

JakeHartnell commented:
For the `cw-curves` package it may be worth looking and the Wynd DAO implementation: https://github.com/wynddao/wynddao/blob/main/packages/utils/src/curve.rs

Probably some useful stuff in there. Would be nice to have a solid curves package.

JakeHartnell committed:
Fix cargo file after rebase

JakeHartnell committed:
Remove token-bindings dep

JakeHartnell committed:
dao-abc-factory contract

Creates new cw-abc DAOs.

JakeHartnell committed:
Fix wasm build

JakeHartnell committed:
Basic Integration Tests

JakeHartnell committed:
Fix clippy errors and test errors

JakeHartnell committed:
Bump nightly version in CI

JakeHartnell committed:
Remove unused deps

JakeHartnell committed:
Clean up small TODO

Commit:
Post rebase fixups

ismellike commented on contracts/external/cw-abc/src/contract.rs:
```diff
@@ -0,0 +1,260 @@
+    // Initialize owner to sender
+    cw_ownable::initialize_owner(deps.storage, deps.api, Some(info.sender.as_str()))?;
+
+    // Tnstantiate cw-token-factory-issuer contract
```
Spelling of Instantiate

ismellike commented on contracts/external/cw-abc/src/msg.rs:
```diff
@@ -0,0 +1,185 @@
+    /// Update the hatch phase configuration
+    Hatch {
+        contribution_limits: Option<MinMax>,
+        // TODO what is the minimum used for?
```
I think paired with an expiration, the minimum could be used to determine if the ABC Hatch phase was successful. 

If expired and the minimum was not met, then the ABC goes into the Closed phase. If expired with the minimum met, then the ABC goes into the Open phase.

ismellike commented on contracts/external/cw-abc/src/commands.rs:
```diff
@@ -0,0 +1,571 @@
+    })];
+
+    // Send funding to fee recipient
+    if funded > Uint128::zero() {
```
This would start to lose state of what's actually in the funding pool.

ismellike committed:
ABC Improvements (#818)

* Make hatcher_allowlist a map

Also cleans up state when going from Hatch -> Open
Various spell-checking fixes

* Allow existing tokens to be used with ABC's

I think it's important for an existing token to be supported by an ABC.
Token DAO's will be able to set up liquidity more easily by simply allowing mint & burn functionality after instantiating the ABC while holding ownership of it.

*Also adds a query for hatcher allowlist
*fixes spell checks I ran into
*moves TokenInfo from dao-voting-token-staked to dao-interface for use in cw-abc

* Complete funding pool logic

Renamed fees_recipient to funding_pool_forwarding for better naming
Made funding_pool_forwarding optional
Allow updating the funding_pool_forwarding by owner
Allow withdrawing from the funding pool by owner

* Implement a circuit breaker

Adds IS_PAUSED state that is checked in execute entry
Also fixes naming in commands

* Add query for initial supply at ABC creation

* Allow donation into the reserve pool

Also fixes donation state being lost after multiple donations

* Move cw-abc curves to its own package cw-curves

* Improve validation for max_supply

* Support DAO membership in hatcher allowlist

Could use a test here
Also fixes some warnings and flag for cw-std

* Clippy fix

* Only clone decimals from curve_state

also reorg buy command to be closer to sell command for readability

* QoL improvements

Derive copy on DecimalPlaces
Catch OverflowError directly in ContractError
Do not set initial supply as supply - will further remove allowing initial supplies

* Revert support for initial supply

* Fix clippy unnecessary clones

* Donate only into the funding pool

* Add a test for donate and withdraw from funding pool

* Add test for dao hatchers + update schemas

Also removes unused deps from cw-abc which now live in cw-curves

* Update Cargo.toml

* Buy & Sell Quotes

Let users know the curve state and returned amount after x payment

* Complete DAO hatchers w/ priority queue

Disallow selling in the hatch phase
Do not validate max_contribution against max_raise

* Fix priority queue ordering on ties

also clippy fixes

* Couple more clippy fixes

ismellike committed:
Fix issue with instantiating cw-abc with hatchers (#830)

allow self to call method

JakeHartnell commented:
Closing in favor of #926
